File: php/tests/Migrations/ViewGuardTest.php
<?php

declare(strict_types=1);

namespace SimpleOrm\Tests\Migrations;

use PDO;
use PHPUnit\Framework\Attributes\Test;
use PHPUnit\Framework\TestCase;
use SimpleOrm\Dialect\SqliteDialect;
use SimpleOrm\Errors\SimpleOrmException;
use SimpleOrm\Metadata\EntityMapLoader;
use SimpleOrm\Migrations\MigrationRunner;
use SimpleOrm\Migrations\MigrationSet;
use SimpleOrm\Migrations\MigrationStep;
use SimpleOrm\Migrations\MigrationVersion;
use SimpleOrm\Migrations\SchemaSync;
use SimpleOrm\Migrations\SqlVersion;
use SimpleOrm\Migrations\SqlVersionStep;
use SimpleOrm\Migrations\VersionBuilder;
use SimpleOrm\Migrations\ViewActions;
use SimpleOrm\Migrations\ViewMigration;
use SimpleOrm\Tests\Migrations\Fixtures\Guard\GuardedTotals;
use SimpleOrm\Tests\Support\TempDatabase;

/**
 * ADR-0017 add.1 (owner): views get adjusted outside the code in urgencies, so
 * a view step's `expectDefinition` guard compares the live definition against
 * the expected previous one before applying — match applies, drift refuses
 * with `MIG-012`, and only `--force` (`$allowViewDrift`) recreates over the
 * drift (with notice). Mirrors `dotnet/tests/SimpleOrm.Tests/ViewGuardTests.cs`.
 */
final class ViewGuardTest extends TestCase
{
    // Public: read from the anonymous ViewMigration step below, a distinct class scope.
    public const string V1_DDL = 'create view guarded_totals as select 1 as answer';

    public const string V2_DDL = 'create view guarded_totals as select 2 as answer';

    private TempDatabase $database;

    private SqliteDialect $dialect;

    private PDO $connection;

    private EntityMapLoader $maps;

    protected function setUp(): void
    {
        $this->database = TempDatabase::create();
        $this->dialect = new SqliteDialect();
        $this->connection = $this->dialect->createConnection($this->database->connectionString());
        $this->maps = new EntityMapLoader();
    }

    protected function tearDown(): void
    {
        unset($this->connection);
        $this->database->delete();
    }

    private function v1(): SqlVersion
    {
        return new SqlVersion(1, new SqlVersionStep('guarded_totals', 'create', [self::V1_DDL], ['drop view guarded_totals']));
    }

    /** The guarded change step (ADR-0017 add.1): hand-written `down()` carries its own guard on `V2_DDL`, the same way `action()` guards `V1_DDL`. */
    private function changeStep(): MigrationStep
    {
        return new class extends ViewMigration {
            public function version(): int
            {
                return 2;
            }

            public function description(): string
            {
                return 'ChangeGuarded';
            }

            public function entityClass(): string
            {
                return GuardedTotals::class;
            }

            public function action(ViewActions $actions): void
            {
                $actions->expectDefinition(ViewGuardTest::V1_DDL);
                $actions->sql('drop view if exists guarded_totals');
                $actions->sql(ViewGuardTest::V2_DDL);
            }

            public function down(ViewActions $actions): void
            {
                $actions->expectDefinition(ViewGuardTest::V2_DDL);
                $actions->sql('drop view if exists guarded_totals');
                $actions->sql(ViewGuardTest::V1_DDL);
            }
        };
    }

    private function v2(): MigrationVersion
    {
        $step = $this->changeStep();

        return new class($step) extends MigrationVersion {
            public function __construct(private readonly MigrationStep $step)
            {
            }

            public function version(): int
            {
                return 2;
            }

            public function compose(VersionBuilder $version): void
            {
                $version->apply($this->step);
            }
        };
    }

    private function runner(): MigrationRunner
    {
        return new MigrationRunner($this->connection, $this->dialect, $this->maps, MigrationSet::of($this->v1(), $this->v2()));
    }

    private function answer(): int
    {
        return (int) $this->connection->query('select answer from guarded_totals')->fetchColumn();
    }

    #[Test]
    public function matching_previous_definition_applies(): void
    {
        $runner = $this->runner();
        self::assertSame(2, $runner->migrate());
        self::assertSame(2, $this->answer());
    }

    #[Test]
    public function outside_drift_refuses_with_mig012_and_applies_nothing(): void
    {
        $only1 = new MigrationRunner($this->connection, $this->dialect, $this->maps, MigrationSet::of($this->v1()));
        $only1->migrate();

        // The urgency hotfix: the view is patched directly in the database.
        SchemaSync::apply($this->connection, ['drop view guarded_totals', 'create view guarded_totals as select 99 as answer']);

        $runner = $this->runner();
        try {
            $runner->migrate();
            self::fail('expected MIG-012');
        } catch (SimpleOrmException $exception) {
            self::assertSame('MIG-012', $exception->errorCode);
        }

        // The whole run rolled back: the hotfixed definition is untouched.
        self::assertSame(99, $this->answer());
    }

    #[Test]
    public function force_recreates_over_the_drift_and_notifies(): void
    {
        $only1 = new MigrationRunner($this->connection, $this->dialect, $this->maps, MigrationSet::of($this->v1()));
        $only1->migrate();
        SchemaSync::apply($this->connection, ['drop view guarded_totals', 'create view guarded_totals as select 99 as answer']);

        $notices = [];
        $runner = $this->runner();
        $applied = $runner->migrate(true, function (string $notice) use (&$notices): void {
            $notices[] = $notice;
        });

        self::assertSame(1, $applied);
        self::assertCount(1, $notices);
        self::assertStringContainsString('guarded_totals', $notices[0]);
        self::assertSame(2, $this->answer());
    }

    #[Test]
    public function down_is_guarded_the_same_way(): void
    {
        $runner = $this->runner();
        $runner->migrate();

        SchemaSync::apply($this->connection, ['drop view guarded_totals', 'create view guarded_totals as select 99 as answer']);
        try {
            $runner->migrateDown(1);
            self::fail('expected MIG-012');
        } catch (SimpleOrmException $exception) {
            self::assertSame('MIG-012', $exception->errorCode);
        }

        self::assertSame(1, $runner->migrateDown(1, true));
        self::assertSame(1, $this->answer());
    }
}

File: go/orm/migrations/view_guard_test.go
package migrations_test

import (
	"context"
	"reflect"
	"testing"

	"github.com/kosmassp/SimpleOrm/go/orm/internal/core"
	"github.com/kosmassp/SimpleOrm/go/orm/internal/metadata"
	"github.com/kosmassp/SimpleOrm/go/orm/migrations"
	"github.com/kosmassp/SimpleOrm/go/orm/sqlite"
)

// ADR-0017 add.1 (owner): views get adjusted outside the code in urgencies,
// so a view step's ExpectDefinition guard compares the live definition
// against the expected previous one before applying — match applies, drift
// refuses with MIG-012, and only --force recreates over the drift (with
// notice). Mirrors ViewGuardTests.cs.

type GuardedTotals struct {
	Answer int64
}

func guardedTotalsMap() *core.EntityMap {
	t := reflect.TypeFor[GuardedTotals]()
	return core.NewEntityMap(
		t, core.RelationView, "guarded_totals", "", "select 1 as answer", nil,
		[]*core.PropertyMap{property(t, "Answer", "answer", core.TypeInt64)}, core.KeyNone, nil, nil)
}

func guardedTotalsLoader() *metadata.Loader {
	options := (&metadata.Options{}).Register(explicitMap{entityType: reflect.TypeFor[GuardedTotals](), m: guardedTotalsMap()})
	return metadata.NewLoader(options)
}

const guardV1DDL = "create view guarded_totals as select 1 as answer"

type V1301_CreateGuarded struct {
	migrations.ViewMigration[GuardedTotals]
}

func (V1301_CreateGuarded) Action(a *migrations.ViewActions) { a.SQL(guardV1DDL) }

type V1301 struct{}

func (V1301) Compose(v *migrations.VersionBuilder) { v.Apply(V1301_CreateGuarded{}) }

type V1302_ChangeGuarded struct {
	migrations.ViewMigration[GuardedTotals]
}

const guardV2DDL = "create view guarded_totals as select 2 as answer"

func (V1302_ChangeGuarded) Action(a *migrations.ViewActions) {
	a.ExpectDefinition(guardV1DDL)
	a.SQL("drop view if exists guarded_totals")
	a.SQL(guardV2DDL)
}

// Down is the manual override, mirroring the guard in reverse (ADR-0017 add.1):
// a rollback must not silently overwrite an outside hotfix either.
func (V1302_ChangeGuarded) Down(a *migrations.ViewActions) {
	a.ExpectDefinition(guardV2DDL)
	a.SQL("drop view if exists guarded_totals")
	a.SQL(guardV1DDL)
}

type V1302 struct{}

func (V1302) Compose(v *migrations.VersionBuilder) { v.Apply(V1302_ChangeGuarded{}) }

func guardedPlan(t *testing.T) *migrations.Set {
	t.Helper()
	set, err := migrations.NewSet(V1301{}, V1302{})
	if err != nil {
		t.Fatal(err)
	}
	return set
}

func TestViewGuard_MatchingPreviousDefinitionApplies(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	runner := migrations.NewRunner(pool, sqlite.New(), guardedTotalsLoader(), guardedPlan(t), nil)

	applied, err := runner.Migrate(ctx, migrations.RunOptions{})
	if err != nil {
		t.Fatal(err)
	}
	if applied != 2 {
		t.Fatalf("expected 2 applied versions, got %d", applied)
	}
	var answer int64
	if err := pool.QueryRowContext(ctx, "select answer from guarded_totals").Scan(&answer); err != nil {
		t.Fatal(err)
	}
	if answer != 2 {
		t.Fatalf("expected the V2 definition to apply, got answer=%d", answer)
	}
}

func TestViewGuard_OutsideDriftRefusesAndAppliesNothing(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	loader := guardedTotalsLoader()

	v1Only, err := migrations.NewSet(V1301{})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := migrations.NewRunner(pool, sqlite.New(), loader, v1Only, nil).Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}

	conn, err := pool.Conn(ctx)
	if err != nil {
		t.Fatal(err)
	}
	defer conn.Close()
	if err := migrations.ApplySync(ctx, conn,
		[]string{"drop view guarded_totals", "create view guarded_totals as select 99 as answer"}); err != nil {
		t.Fatal(err)
	}

	runner := migrations.NewRunner(pool, sqlite.New(), loader, guardedPlan(t), nil)
	_, err = runner.Migrate(ctx, migrations.RunOptions{})
	if core.CodeOf(err) != "MIG-012" {
		t.Fatalf("expected MIG-012, got %v", err)
	}

	// The whole run rolled back: the hotfixed definition is untouched.
	var answer int64
	if err := pool.QueryRowContext(ctx, "select answer from guarded_totals").Scan(&answer); err != nil {
		t.Fatal(err)
	}
	if answer != 99 {
		t.Fatalf("expected the hotfixed definition to survive the refused run, got answer=%d", answer)
	}
}

func TestViewGuard_ForceRecreatesOverDriftAndNotifies(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	loader := guardedTotalsLoader()

	v1Only, err := migrations.NewSet(V1301{})
	if err != nil {
		t.Fatal(err)
	}
	if _, err := migrations.NewRunner(pool, sqlite.New(), loader, v1Only, nil).Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}
	conn, err := pool.Conn(ctx)
	if err != nil {
		t.Fatal(err)
	}
	defer conn.Close()
	if err := migrations.ApplySync(ctx, conn,
		[]string{"drop view guarded_totals", "create view guarded_totals as select 99 as answer"}); err != nil {
		t.Fatal(err)
	}

	var notices []string
	runner := migrations.NewRunner(pool, sqlite.New(), loader, guardedPlan(t), nil)
	applied, err := runner.Migrate(ctx, migrations.RunOptions{
		AllowViewDrift: true, Notify: func(n string) { notices = append(notices, n) },
	})
	if err != nil {
		t.Fatal(err)
	}
	if applied != 1 {
		t.Fatalf("expected 1 applied version (V1301 already recorded), got %d", applied)
	}
	if len(notices) != 1 {
		t.Fatalf("expected exactly one drift notice, got %v", notices)
	}

	var answer int64
	if err := pool.QueryRowContext(ctx, "select answer from guarded_totals").Scan(&answer); err != nil {
		t.Fatal(err)
	}
	if answer != 2 {
		t.Fatalf("expected the code's definition recreated, got answer=%d", answer)
	}
}

func TestViewGuard_DownIsGuardedTheSameWay(t *testing.T) {
	ctx := context.Background()
	pool := openMigrationsPool(t)
	loader := guardedTotalsLoader()
	runner := migrations.NewRunner(pool, sqlite.New(), loader, guardedPlan(t), nil)
	if _, err := runner.Migrate(ctx, migrations.RunOptions{}); err != nil {
		t.Fatal(err)
	}

	conn, err := pool.Conn(ctx)
	if err != nil {
		t.Fatal(err)
	}
	defer conn.Close()
	if err := migrations.ApplySync(ctx, conn,
		[]string{"drop view guarded_totals", "create view guarded_totals as select 99 as answer"}); err != nil {
		t.Fatal(err)
	}

	_, err = runner.MigrateDown(ctx, 1301, migrations.RunOptions{})
	if core.CodeOf(err) != "MIG-012" {
		t.Fatalf("expected MIG-012 on the way down too, got %v", err)
	}

	reverted, err := runner.MigrateDown(ctx, 1301, migrations.RunOptions{AllowViewDrift: true})
	if err != nil {
		t.Fatal(err)
	}
	if reverted != 1 {
		t.Fatalf("expected 1 reverted version, got %d", reverted)
	}
	var answer int64
	if err := pool.QueryRowContext(ctx, "select answer from guarded_totals").Scan(&answer); err != nil {
		t.Fatal(err)
	}
	if answer != 1 {
		t.Fatalf("expected the V1 definition restored, got answer=%d", answer)
	}
}
